### PR TITLE
[FLINK-3547] add support for streaming filter, project, and union in the Table API

### PR DESCRIPTION
**Pull request by vasia**
This PR adds:

- Java and Scala stream translators
- methods to convert a table to a DataStream and vice versa
- streaming rules to FlinkRuleSets
- DataStream rules for calc, scan, and union
- tests and streaming test utils

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/datastream/DataStreamCalc.scala [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+import org.apache.flink.streaming.api.datastream.DataStream
+
+import scala.collection.JavaConversions._
+import scala.collection.JavaConverters._
+
+/**
+  * Flink RelNode which matches along with FlatMapOperator.
+  *
+  */
+class DataStreamCalc(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/datastream/DataStreamCalc.scala]: This class shares a lot of code with `DataSetCalc`. Can we extract common parts such generation of the FlatMapFunction and the name of the operator into a common `FlinkCalc` trait or so?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/datastream/DataStreamRel.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    *                     type Row. If the expected type is Tuple2, the operator will return
+    *                     a Tuple2 if possible. Row otherwise.
+    * @return DataStream of type expectedType or RowTypeInfo
+    */
+  def translateToPlan(
+      config: TableConfig,
+      expectedType: Option[TypeInformation[Any]] = None)
+    : DataStream[Any]
+
+  private[flink] def getExpressionString(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/datastream/DataStreamRel.scala]: Same as `DataSetRel.getExpressionString()`? Move to a `FlinkRel` trait?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/schema/DataStreamTable.scala [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+import org.apache.calcite.rel.`type`.{RelDataType, RelDataTypeFactory}
+import org.apache.calcite.schema.impl.AbstractTable
+import org.apache.calcite.sql.`type`.SqlTypeName
+import org.apache.flink.api.common.typeinfo.AtomicType
+import org.apache.flink.api.common.typeutils.CompositeType
+import org.apache.flink.api.table.typeutils.TypeConverter
+import org.apache.flink.streaming.api.datastream.DataStream
+
+class DataStreamTable[T](
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/schema/DataStreamTable.scala]: Move code that is shared with `DataSetTable` into a common abstract `FlinkTable` class?

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/table/streaming/test/FilterITCase.scala [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.collection.JavaConversions._
+import org.junit.Test
+import org.junit.Assert._
+import org.apache.flink.api.scala.table.streaming.test.utils.StreamITCase
+import org.apache.flink.api.scala.table.streaming.test.utils.StreamTestData
+
+class FilterITCase extends StreamingMultipleProgramsTestBase {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/table/streaming/test/FilterITCase.scala]: Add a test that checks correct exception if predicate attribute does not exist.

**Comment by fhueske** [comment]
I think we should try to share more code between the DataSet and DataStream translation. Otherwise it looks good. 

We should also rework the tests and extract tests that check the Table API parts that are common for DataSet and DataStream. I wouldn't do that in this PR though.

**Comment by vasia** [comment]
Thanks for the review! I'll see how we can share the common parts between DataSet and DataStream translation. I will also open a JIRA for reworking the tests.

**Comment by vasia** [comment]
I have updated the PR. Thanks!

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/FlinkCalc.scala [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+  }
+
+  private[flink] def calcToString(
+      calcProgram: RexProgram,
+      expression: (RexNode, List[String], Option[List[RexNode]]) => String) = {
+
+    val conditionStr = conditionToString(calcProgram, expression)
+    val selectionStr = selectionToString(calcProgram, expression)
+
+    s"Calc(${if (calcProgram.getCondition != null) {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/FlinkCalc.scala]: Reuse `calcOpName` here?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/FlinkCalc.scala [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+import org.apache.flink.api.table.codegen.{GeneratedFunction, CodeGenerator}
+import org.apache.flink.api.table.runtime.FlatMapRunner
+import org.apache.flink.api.table.typeutils.TypeConverter._
+
+import scala.collection.JavaConversions._
+import scala.collection.JavaConverters._
+
+trait FlinkCalc {
+
+  def functionBody(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/FlinkCalc.scala]: make as well `private[flink]`?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetRel.scala [hunk 7]
```diff
@@ -20,15 +20,15 @@ package org.apache.flink.api.table.plan.nodes.dataset
...
 import org.apache.calcite.sql.`type`.SqlTypeName
 import org.apache.flink.api.common.typeinfo.TypeInformation
 import org.apache.flink.api.java.DataSet
 import org.apache.flink.api.table.TableConfig
+import org.apache.flink.api.table.plan.nodes.FlinkRel
 
 import scala.collection.JavaConversions._
 
-trait DataSetRel extends RelNode {
+trait DataSetRel extends RelNode with FlinkRel{
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetRel.scala]: space after `FlinkRel`

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/datastream/DataStreamCalc.scala [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    cluster: RelOptCluster,
+    traitSet: RelTraitSet,
+    input: RelNode,
+    rowType: RelDataType,
+    calcProgram: RexProgram,
+    opName: String,
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/datastream/DataStreamCalc.scala]: `opName` is no longer used and can be removed.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/datastream/DataStreamUnion.scala [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  * Flink RelNode which matches along with Union.
+  *
+  */
+class DataStreamUnion(
+    cluster: RelOptCluster,
+    traitSet: RelTraitSet,
+    left: RelNode,
+    right: RelNode,
+    rowType: RelDataType,
+    opName: String)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/datastream/DataStreamUnion.scala]: `DataSetUnion` was updated to not use the `opName` anymore. Would be good to reflect these changes also in `DataStreamUnion`.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/datastream/DataStreamScanRule.scala [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    classOf[LogicalTableScan],
+    Convention.NONE,
+    DataStreamConvention.INSTANCE,
+    "DataStreamScanRule")
+{
+
+  def convert(rel: RelNode): RelNode = {
+    val scan: LogicalTableScan = rel.asInstanceOf[LogicalTableScan]
+    val traitSet: RelTraitSet = rel.getTraitSet.replace(DataStreamConvention.INSTANCE)
+
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/datastream/DataStreamScanRule.scala]: remove 1 new line

**Comment by fhueske** [comment]
Proposed a few refactorings. Looks good otherwise.

**Comment by vasia** [comment]
Thanks. I've addressed you comments!

**Comment by fhueske** [comment]
Thanks for the update! +1 to merge :-)

**Comment by vasia** [comment]
I've rebased on master and will merge once travis turns green :)

**Comment by vasia** [comment]
merging this

**Pull request closed without merging**